### PR TITLE
initialize conda

### PR DESCRIPTION
**Pull request by volodymyrss**

### CODE REVIEW

**Comment by dsavchenko** [comment]
This module `container.py` isn't actually used in `deploy.py`
Is it used anywhere besides the `nb2worker` cli command?

Code restructuring probably needed. The Dockerfile creation is in the function in `deploy.py`

**Comment by volodymyrss** [comment]
> This module `container.py` isn't actually used in `deploy.py` Is it used anywhere besides the `nb2worker` cli command?
> 
> Code restructuring probably needed. The Dockerfile creation is in the function in `deploy.py`

hm, ok. Well. this change here is still needed.

The other part is written by you I think, could you add similar change there?

**Comment by dsavchenko** [comment]
> The other part is written by you I think, could you add similar change there?

It was adapted by me, not written from scratch. I didn't even notice the existence of this `container.py`, as the alternative Dockerfile generation was already there. So we should definitely remove this redundancy, probably with other Issue/PR. 
Corresponding functions in `deploy.py` have now additional features, do they needed for the use case of `nb2worker`?

Anyway, OK, I will do changes in that part.

**Comment by volodymyrss** [comment]
> > The other part is written by you I think, could you add similar change there?
> 
> It was adapted by me, not written from scratch. I didn't even notice the existence of this `container.py`, as the alternative Dockerfile generation was already there. So we should definitely remove this redundancy, probably with other Issue/PR. Corresponding functions in `deploy.py` have now additional features, do they needed for the use case of `nb2worker`?

Think there was some important reason to have it separately, even if not a very good one. Container generation should be in "container", there should be an option to keep original container as it is now, etc. By the way we might also want to just run the original container without rebuilding, but it takes extra effort since we need to potentially scan it for notebooks.

> Anyway, OK, I will do changes in that part.

ok! Just to see euclid and fermi work.

**Comment by dsavchenko** [comment]
> By the way we might also want to just run the original container without rebuilding, but it takes extra effort since we need to potentially scan it for notebooks.

We have an option to base the container on the renku one (still adding nb2workflow and running it, though)

**Comment by volodymyrss** [comment]
> > By the way we might also want to just run the original container without rebuilding, but it takes extra effort since we need to potentially scan it for notebooks.
> 
> We have an option to base the container on the renku one (still adding nb2workflow and running it, though)

Yeah but only in one of duplicate implementations. But just to keep in mind that this is important.

File: nb2workflow/container.py [hunk 1]
```diff
@@ -55,10 +55,12 @@ def prepare_image(repo_source, from_image, service=True, nb2w_path=None, runpref
         dockerfile.append("ADD $REPO_PATH/pip.conf /etc/pip/pip.conf")
 
     dockerfile.append("ADD $REPO_PATH/requirements.txt /requirements.txt")
+    dockerfile.append("ADD $REPO_PATH/environment.yml /environment.yml")
     dockerfile.append("RUN {} pip install --upgrade pip".format(runprefix))
-    # ,repo_hash))
-    dockerfile.append(
-        "RUN {} pip install -r /requirements.txt --upgrade".format(runprefix))
+   
+    dockerfile.append(('RUN {} conda env update -f /environment.yml && '
```
Comment (dsavchenko) [inline nb2workflow/container.py:61]: This won't work in all cases. Conda should be installed and initialised first.

**Comment by volodymyrss** [comment]
althought it's you @dsavchenko who added much of the code, please accept the PR if you are ready, it looks ok for me

**Comment by dsavchenko** [comment]
I changed a lot. Now using mamba for conda environment (it's lot better in resolving dependencies). 
It's now deployed at staging, seems to be working well.
@volodymyrss  I let you see the changes one more time

**Comment by volodymyrss** [comment]
> althought it's you @dsavchenko who added much of the code, please accept the PR if you are ready, it looks ok for me

I think it looks good, thanks!
The request still fails but for a different reason probably quite unrelated to this issue.

**Pull request merged**